chains/evm/contracts/ccvs/VersionedVerifierResolver.sol

```diff
@@ -0,0 +1,140 @@
+
+import {EnumerableSet} from "@openzeppelin/contracts@5.0.2/utils/structs/EnumerableSet.sol";
+
+/// @notice Resolves and returns the appropriate verifier contract for the given outbound / inbound traffic.
+/// @dev On source, the destChainSelector of a message is used to determine the verifier implementation to apply.
+/// On destination, we must use the verifier version was applied on source, parsing this version from the ccvData.
+contract VersionedVerifierResolver is ICrossChainVerifierResolver, ITypeAndVersion, Ownable2StepMsgSender {
+  using EnumerableSet for EnumerableSet.UintSet;
+  using EnumerableSet for EnumerableSet.Bytes32Set;
+
+  error InvalidCCVDataLength();
+  error InvalidDestChainSelector(uint64 destChainSelector);
+  error InvalidVersion(bytes4 version);
+
+  event InboundImplementationRemoved(bytes4 version);
+  event OutboundImplementationRemoved(uint64 destChainSelector);
+  event InboundImplementationUpdated(bytes4 version, address prevImpl, address newImpl);
+  event OutboundImplementationUpdated(uint64 destChainSelector, address prevImpl, address newImpl);
+
+  struct InboundImplementationArgs {
+    bytes4 version; // ────╮ Verifier version.
+    address verifier; // ──╯ Address of the verifier contract.
+  }
+
+  struct OutboundImplementationArgs {
+    uint64 destChainSelector; // ──╮ Destination chain selector.
+    address verifier; // ──────────╯ Address of the verifier contract.
+  }
+
+  string public constant override typeAndVersion = "VersionedVerifierResolver 1.7.0-dev";
+
+  /// @notice maps verifier versions to their implementation addresses, applied to inbound traffic.
+  mapping(bytes4 version => address verifier) private s_versionToInboundImplementation;
+  /// @notice all supported verifier versions.
+  EnumerableSet.Bytes32Set private s_supportedVerifierVersions;
+  /// @notice maps destination chain selectors to their implementation addresses, applied to outbound traffic.
+  mapping(uint64 destChainSelector => address version) private s_destChainToOutboundImplementation;
+  /// @notice all supported destination chains.
+  EnumerableSet.UintSet private s_supportedDestChains;
+
+  /// @inheritdoc ICrossChainVerifierResolver
+  function getInboundImplementation(
+    bytes calldata ccvData
+  ) external view returns (address) {
+    if (ccvData.length < 4) {
+      revert InvalidCCVDataLength();
+    }
+    return s_versionToInboundImplementation[bytes4(ccvData[:4])];
+  }
+
+  /// @notice Returns the verifier contract for a given version.
+  /// @param version The version of the verifier contract.
+  /// @return verifierAddress The address of the verifier contract.
+  function getInboundImplementationForVersion(
+    bytes4 version
+  ) external view returns (address) {
+    return s_versionToInboundImplementation[version];
+  }
+
+  /// @notice Returns all supported verifier versions.
+  function getSupportedVerifierVersions() external view returns (bytes4[] memory) {
```

Comment:
Maybe return InboundImplementationArgs/outbound args in these functions so you have all info and don't need to do e.g. 20 calls to get the verifier for each of the 20 versions